### ISSUE TITLE
Add support for standard SQL MERGE .. WHEN MATCHED [ AND .. ] THEN DELETE

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10052]
The SQL standard and a variety of dialects support one or many `WHEN MATCHED [ AND .. ] THEN DELETE` clauses in the `MERGE` statement. We already support Oracle's `DELETE WHERE` syntax.

**Native support:**

- [x] DB2
- [x] Derby
- [x] H2
- [x] HSQLDB (only one `WHEN MATCHED .. UPDATE` and one `WHEN MATCHED .. DELETE` clause are allowed)
- [x] SQL Server
- [x] Sybase SQL Anywhere
- [x] Teradata

**Emulation through Oracle style `WHERE`**:

- [x] Oracle

----
See also:  #7291

**Comment by lukaseder** [comment]
In standard SQL, it is possible to write:

```sql
MERGE INTO t USING s ON p WHEN MATCHED THEN DELETE;
MERGE INTO t USING s ON p WHEN MATCHED AND q THEN DELETE;
```

This isn't possible in Oracle. A `DELETE` clause always requires an `UPDATE` clause as well. If we know the table's columns, then we can apply a trick again:

```
MERGE INTO t USING s ON p WHEN MATCHED THEN UPDATE j = j DELETE WHERE 1 = 1;
MERGE INTO t USING s ON p WHEN MATCHED THEN UPDATE j = j WHERE q DELETE WHERE q;
```

It may be tempting to think that we can add a `WHERE 1 = 0` clause to `UPDATE`, but we cannot. Only rows can be deleted that were updated before. Very weird feature.

The predicate `q` (if present) has to be repeated to get the correct update count and locking behaviour.

We don't care about the fact that this emulation will trigger `UPDATE` triggers as well. I don't think there's a way to prevent that.

While the above query could perhaps be transformed into an ordinary `DELETE` statement, there might also be a `WHEN NOT MATCHED` clause, in case of which we'd have to revert to this emulation anyway.

**Comment by lukaseder** [comment]
> It may be tempting to think that we can add a `WHERE 1 = 0` clause to `UPDATE`, but we cannot. Only rows can be deleted that were updated before. Very weird feature.

This actually means we cannot just "hi-jack" any arbitrary `UPDATE` expression. In the presence of any `DELETE` expressions, we'll have to make sure that all the rows that are deleted are also updated with such a `j = j` expression first. This is quite a laborious emulation, but doable. For example

```sql
-- Standard SQL
MERGE INTO t USING s ON p 
WHEN MATCHED AND p1 THEN DELETE
WHEN MATCHED AND p2 THEN UPDATE SET a = 1
WHEN MATCHED AND p3 THEN DELETE
WHEN MATCHED AND p4 THEN UPDATE SET b = 2

-- Oracle SQL
MERGE INTO t USING s ON p
WHEN MATCHED THEN UPDATE SET
  a = CASE 
      WHEN p1 THEN a
      WHEN p2 THEN 1 -- Actual value
      WHEN p3 THEN a
      WHEN p4 THEN a
              ELSE a
    END, 
  b = CASE
      WHEN p1 THEN b
      WHEN p2 THEN b
      WHEN p3 THEN b
      WHEN p4 THEN 2 -- Actual value
              ELSE b
    END
  WHERE p1 OR p2 OR p3 OR p4
  DELETE WHERE p1 OR (NOT p1 AND NOT p2 AND p3) -- Negation of previous predicates!
```

In theory, with some more sophisticated SQL transformation (see #10054), the case expressions for this example could be simplified to become this:

```sql
-- Oracle SQL
MERGE INTO t USING s ON p
WHEN MATCHED THEN UPDATE SET
  a = CASE WHEN NOT p1 AND p2 THEN 1 ELSE a END, 
  b = CASE WHEN NOT p1 AND NOT p2 AND NOT p3 AND p4 THEN 2 ELSE b END
  WHERE p1 OR p2 OR p3 OR p4
  DELETE WHERE p1 OR (NOT p1 AND NOT p2 AND p3) -- Negation of previous predicates!
```

But that is quite difficult to get right.

**Comment by lukaseder** [comment]
The additional Oracle emulations will be implemented through #7291.
